src/main/java/com/fakng/fakngagrgtr/parser/apple/AppleParser.java

```diff
@@ -0,0 +1,131 @@
+import net.minidev.json.parser.JSONParser;
+import net.minidev.json.parser.ParseException;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.PostConstruct;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Component
+public class AppleParser extends HtmlParser {
+
+    private static final String URL_FOR_VACANCY = "https://jobs.apple.com/en-us/details/";
+    private static final String ELEMENT_NAME_WITH_PAGE_COUNT = "frmPagination";
+    private static final String ELEMENT_TAG_WITH_PAGE_COUNT = "span";
+    private static final String PATH_TO_VACANCIES = "searchResults";
+    private static final String PATH_TO_FIELD_WITH_JSON_BODY = ".//script";
+    private static final JSONParser jsonParser = new JSONParser(JSONParser.MODE_JSON_SIMPLE);
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+
+    public AppleParser(WebClient htmlWebClient,
+                       CompanyRepository companyRepository,
+                       LocationProcessor locationProcessor,
+                       @Value("${url.apple}") String url) {
+        super(htmlWebClient, companyRepository, locationProcessor);
+        this.url = url;
+    }
+
+    @Override
+    protected String getCompanyName() {
+        return "Apple";
+    }
+
+    @PostConstruct
+    public void init() {
+        initBase();
+    }
+
+    @Override
+    protected List<Vacancy> getAllVacancies() throws Exception {
+        HtmlPage htmlPage = getResponseByPage(0);
+        int countPages = getCountPages(htmlPage);
+        List<Vacancy> allVacancies = new ArrayList<>(getVacanciesFromDoc(htmlPage));
+        for (int i = 1; i < countPages; i++) {
+            htmlPage = getResponseByPage(i);
+            allVacancies.addAll(getVacanciesFromDoc(htmlPage));
+        }
+        return allVacancies;
+    }
+
+    private HtmlPage getResponseByPage(int numberOfPage) throws IOException {
+        return downloadPage(String.format(url, numberOfPage));
+    }
+
+    private int getCountPages(HtmlPage htmlPage) {
+        return Integer.parseInt(htmlPage.getElementsByIdAndOrName(ELEMENT_NAME_WITH_PAGE_COUNT).get(0).getElementsByTagName(ELEMENT_TAG_WITH_PAGE_COUNT).get(1).getVisibleText());
+    }
+
+    private List<Vacancy> getVacanciesFromDoc(HtmlPage htmlpage) throws ParseException, JsonProcessingException {
+        String jsonBody = getJsonBody(htmlpage);
+        return getVacanciesFromJsonBody(jsonBody);
+    }
+
+    private String getJsonBody(HtmlPage htmlPage) {
+        String textContent = ((HtmlScript) htmlPage.getByXPath(PATH_TO_FIELD_WITH_JSON_BODY).get(1)).getTextContent();
+        return textContent.substring(textContent.indexOf('{'), textContent.lastIndexOf('}') + 1);
+    }
+
+    private List<Vacancy> getVacanciesFromJsonBody(String json) throws ParseException, JsonProcessingException {
```

Comment:
Okay, just to better understand

You're taking json param here, which is String
Then you parse this String and make it JSONObject
Afterwards, you take pass this JSONObject to toVacancyApple method, where it's again casted to String

What's the purpose of the second step?

Reply:
Initially, I will transform the json string into a JSONObject so that it is convenient to get an array of vacancies from there. Otherwise, it would have to be parsed through String, which, in my opinion, would look bad. Unfortunately, I didn't find a method to convert JSONObject to Class, only String to Class.

Reply:
I think ObjectMapper can parse your original JSON to List<VacancyDTO> or VacancyDTO[]. Because as I see the main problem here is about extracting array from the original JSON...

Reply:
Or you can even create an entire new DTO class for parsing this original JSON with ObjectMapper. And then just get your List from there

Reply:
It seems superfluous to me. It will be a huge class with a bunch of fields to get a small part of it. Don't you think so?

Reply:
I guess, you can just ignore most of them. There's no need to write them all. If you write only those fields that you need, ObjectMapper will be smart enough to provide you only with those. Worth trying doing this. If it works really, then it makes things much simpler

Reply:
For some reason, I thought it would be more difficult, and I didn't do it right away. But it turned out to be easier